### PR TITLE
Change `as_setup` and `as_teardown` to instance methods

### PR DESCRIPTION
**Pull request by dstandish**
TLDR: 
* provides a oneline syntax for setting setup / teardown deps
* makes it easy to convert dags to use feature
* provides a mechanism to combine "reusable" taskflow tasks with setup / teardown
* set setup and teardown in the same place you set deps

This provides a number of benefits.

One is now we can do MyTeardown(...).as_teardown(the_setup) instead of having to arrow directly these two.

The other is that I think it can be cognitively easier to set these properties while setting relationships rather than when instantiating.

This can also make it easier to "convert" an existing dag to use setup / teardown.  For example, take a look at [this system test dag](https://github.com/apache/airflow/blob/main/tests/system/providers/amazon/aws/example_dms.py#L407-L429).  Imagine converting that to use setup / teardown.  If you need to set setup / teardown at instantiation, you are doing it far away from where you are actually setting the relationships between the tasks.  But I think sometimes it makes a lot of sense to set those properties at the same time as you set those relationships.  The present PR doesn't force user to do this, but it makes it possible.

Finally, this *also* makes it possible to define reusable, parameterized tasks in taskflow, and use them as setups / teardowns / work tasks depending on context.  If we force you to set it at decorator level, then you must pick and we don't provide an official mechanism to override.

Example of how this can make authoring a bit cleaner. I give two ways supported with the new approach, and for comparison i include the old way you had to do it.

New way 1:
```python
with DAG(
    dag_id="example_setup_teardown",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
) as dag:
    s1 = BashOperator(task_id="setup", bash_command="echo 'Hello from setup'")
    w1 = BashOperator(task_id="normal", bash_command="echo 'I am just a normal task'")
    t1 = BashOperator(
        task_id="teardown", bash_command="echo 'Goodbye from teardown'"
    )
    s1.as_setup() >> w1 >> t1.as_teardown(s1)
```

new way 2:
```python
with DAG(
    dag_id="example_setup_teardown",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
) as dag:
    s1 = BashOperator(task_id="setup", bash_command="echo 'Hello from setup'").as_setup()
    w1 = BashOperator(task_id="normal", bash_command="echo 'I am just a normal task'")
    t1 = BashOperator(
        task_id="teardown", bash_command="echo 'Goodbye from teardown'"
    ).as_teardown(s1)
    s1 >> w1 >> t1
```

Old way:
```python
with DAG(
    dag_id="example_setup_teardown",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
) as dag:
    s1 = BashOperator.as_setup(task_id="setup", bash_command="echo 'Hello from setup'")
    w1 = BashOperator(task_id="normal", bash_command="echo 'I am just a normal task'")
    t1 = BashOperator.as_teardown(
        task_id="teardown", bash_command="echo 'Goodbye from teardown'"
    )
    s1 >> w1 >> t1
    s1 >> t1
```

### CODE REVIEW

**Comment by ephraimbuddy** [comment]
This looks good but what do you think about not having `as_teardown(setuptask)`?

**Comment by dstandish** [comment]
> This looks good but what do you think about not having `as_teardown(setuptask)`?

What don't you like about it?

**Comment by ephraimbuddy** [comment]
> > This looks good but what do you think about not having `as_teardown(setuptask)`?
> 
> What don't you like about it?
> 
> Would you prefer kwargs only?
> 
> Or perhaps a separate method such as `with_setup`? Or `teardown_for`? But just seems a bit redundant.

I was thinking that `with s1 >> t1` should suffix. Makes the syntax similar to what we have with the decorators

**Comment by jedcunningham** [comment]
I lean toward kwarg only, more explicit.

However, I'm not sure I like `as_setup` and `as_teardown` as instance methods. I guess that's more of a personal preference thing though.

Here is why:

```
# makes sense, s1/t1 wrap w1
s1.as_setup() >> w1 >> t1.as_teardown(s1)

# later in the file...
s1 >> w2 >> t1
```

Should I `s1.as_setup()` again? If I don't, what happens when I remove the original w1 lines `as_setup`? Keeping it during task instantiation avoids this problem. I don't find the "easier to convert" thing to be that big of a deal. Taskflow reuse is a valid concern though, but having a setup and non-setup share implementation isn't that difficult to pull off.

My 2c, not strongly held. @uranusjr, what do you think?

File: airflow/models/taskmixin.py [hunk 1]
```diff
@@ -69,6 +70,14 @@ def set_downstream(
         """Set a task or a task list to be directly downstream from the current task."""
         raise NotImplementedError()
 
+    def as_setup(self):
```
Comment (ephraimbuddy) [inline airflow/models/taskmixin.py]: Should we mark this and the other one as an `abstractmethod`? I think with this, we are also going to have people try to use task_group as setup/teardown

Comment (dstandish) [reply airflow/models/taskmixin.py]: reason i didn't do this is i don't necessarily want all subclasses to implement.  e.g. it's not supported for mapped operator, which is a different xcomarg class from PlainXComArg cc @uranusjr if you have thoughts on this PR

File: airflow/example_dags/example_setup_teardown_taskflow.py [hunk 2]
```diff
@@ -47,12 +45,10 @@ def normal():
     @task_group
     def section_1():
         # You can also have setup and teardown tasks at the task group level
-        @setup
         @task
         def my_setup():
             print("I set up")
 
-        @teardown
```
Comment (ephraimbuddy) [inline airflow/example_dags/example_setup_teardown_taskflow.py:55]: Should we create another example dag to demonstrate this usage instead of removing this one?

Comment (dstandish) [reply airflow/example_dags/example_setup_teardown_taskflow.py:55]: yeah you are right

**Comment by dstandish** [comment]
OK @uranusjr @ephraimbuddy @jedcunningham  the scope of this PR expanded a little bit.
I wanted to add validation so that (1) a task can't be both setup and teardown, (2) once you set a task to be teardown, you can't change it to be setup or vice versa, (3) for mapped operator, explicitly forbid users from setting is_teardown or is_setup or on_failure_fail_dagrun, and (4) forbid user from setting `on_failure_fail_dagrun` unless the task is a teardown task.

Clamping down a little more in this way should make the feature a little more user friendly by preventing some usages / configurations that don't make sense or are in any case unsupported.

To accomplish this I changed the three attrs to be properties.  I added them to abstractoperator and I override the setter in mappedoperator to throw an error.

The reason it made sense in abstract operator instead of baseoperator is that in taskflow / xcomarg the iter_references method yields type `Operator` which is either base or mapped operator which both inherit separately from abstractoperator.

So when implementing as_teardown etc for taskflow, we can just set the attrs without inspecting type and let mapped operator throw when it's not supported.

Ready for a look now.  Thank you.

File: tests/models/test_taskmixin.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    # can't change a setup task to a teardown task or vice versa
+    for first, second in [("setup", "teardown"), ("teardown", "setup")]:
+        with dag_maker():
+            s1 = make_task(name="s1", type_=type_)
+            getattr(s1, f"as_{first}")()
+            with pytest.raises(
+                ValueError, match=f"Cannot mark task 's1' as {second}; task is already a {first}."
+            ):
+                getattr(s1, f"as_{second}")()
+                s1.as_teardown()
```
Comment (ephraimbuddy) [inline tests/models/test_taskmixin.py:144]:
```suggestion
```
I don't think we need this?

File: airflow/example_dags/example_setup_teardown_taskflow.py [hunk 4]
```diff
@@ -29,30 +29,56 @@
...
         print("I am just a normal task")
 
+    # since we already marked these as setup / teardown, we just need to make sure they are linked
+    # here we make sure setup and teardown are connected.
+    # if not using `as_teardown` we must arrow them explicitly
+    s = dag_setup()
+    t = dag_teardown()
+    s >> t
+    # and here we add our "work" task a.k.a. normal task.
+    s >> dag_normal_task() >> t
```
Comment (ephraimbuddy) [inline airflow/example_dags/example_setup_teardown_taskflow.py]: I think we should use context manager here

File: airflow/models/abstractoperator.py [hunk 5]
```diff
@@ -149,6 +155,92 @@ def dag_id(self) -> str:
...
+        return self._is_setup
+
+    @is_setup.setter
+    def is_setup(self, value):
+        """
+        Setter for is_setup property.
+
+        :meta private:
+        """
+        if self.is_teardown is True and value is True:
```
Comment (ephraimbuddy) [inline airflow/models/abstractoperator.py:174]:
```suggestion
        if self.is_teardown is True:
```
I think being strict here is better to avoid bad coding from 'copying'

File: airflow/models/abstractoperator.py [hunk 6]
```diff
@@ -149,6 +155,92 @@ def dag_id(self) -> str:
...
+        return self._is_teardown
+
+    @is_teardown.setter
+    def is_teardown(self, value):
+        """
+        Setter for is_teardown property.
+
+        :meta private:
+        """
+        if self.is_setup is True and value is True:
```
Comment (ephraimbuddy) [inline airflow/models/abstractoperator.py:194]:
```suggestion
        if self.is_setup is True:
```

File: airflow/models/abstractoperator.py [hunk 7]
```diff
@@ -149,6 +155,92 @@ def dag_id(self) -> str:
...
+        return self._on_failure_fail_dagrun
+
+    @on_failure_fail_dagrun.setter
+    def on_failure_fail_dagrun(self, value):
+        """
+        Setter for on_failure_fail_dagrun property.
+
+        :meta private:
+        """
+        if value is True and self.is_teardown is not True:
```
Comment (ephraimbuddy) [inline airflow/models/abstractoperator.py:214]:
```suggestion
        if not self.is_teardown:
```

File: tests/models/test_taskmixin.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    # after setting deps, still none are setup / teardown
+    # verify relationships
+    s1 >> w1 >> t1
+    assert cleared_tasks(dag, "w1") == {"w1", "t1"}
+    assert get_task_attr(t1, "is_teardown") is False
+    assert get_task_attr(s1, "is_setup") is False
+    assert get_task_attr(t1, "upstream_task_ids") == {"w1"}
+
+    # now when we use as_teardown, s1 should be setup, t1 should be teardown, and we should have s1 >> t1
+    t1.as_teardown(s1)
```
Comment (ephraimbuddy) [inline tests/models/test_taskmixin.py]: This makes me think we should make `as_teardown` take keyword arg because this reads like we want to make s1 the teardown.

File: airflow/models/abstractoperator.py (hunk 7 above)
Comment (dstandish) [reply airflow/models/abstractoperator.py:214]: need both because it's a default and it gets set to false when reinitialized in certain contexts

File: airflow/models/abstractoperator.py (hunk 6 above)
Comment (dstandish) [reply airflow/models/abstractoperator.py:194]: needed because it's set to false at init and this should be allowed

File: airflow/models/abstractoperator.py (hunk 5 above)
Comment (dstandish) [reply airflow/models/abstractoperator.py:174]: oh i think i see what you mean

File: airflow/example_dags/example_setup_teardown_taskflow.py (hunk 4 above)
Comment (dstandish) [reply airflow/example_dags/example_setup_teardown_taskflow.py]: updated

File: airflow/models/abstractoperator.py (hunk 5 above)
Comment (dstandish) [reply airflow/models/abstractoperator.py:174]: ok i tried it but yeah we can really do this becasue it then it prevents setting it with even the default value at init.  reverted for now. if you have another suggestion then i'm all ears

File: tests/models/test_taskmixin.py (hunk 8 above)
Comment (dstandish) [reply tests/models/test_taskmixin.py]: user *can* use kwargs if they want. do you really think we need to make it kwargs only?

Comment (ephraimbuddy) [reply tests/models/test_taskmixin.py]: No strong opinion

**Review by ephraimbuddy — APPROVED** [review]

File: tests/models/test_taskmixin.py (hunk 3 above)
Comment (ephraimbuddy) [reply tests/models/test_taskmixin.py:144]: What do you think about this?

File: tests/models/test_taskmixin.py (hunk 8 above)
Comment (dstandish) [reply tests/models/test_taskmixin.py]: ok i'll revert it for now (making it kwargs-_optional_) but no problem revisiting if desired

Comment (dstandish) [reply tests/models/test_taskmixin.py]: actually, no, i'll keep it kwargs only for now
